File: python/pov/pov.py
"""Tree POV."""


from __future__ import annotations
from typing import Generator, Optional


class Tree:
    """A Tree which can be reoriented."""

    def __init__(self, label: str, children: Optional[list[Tree]] = None):
        """Initialize a Tree Node."""
        self.label = label
        self.children = children or []
        self.parent = None
        for child in self.children:
            child.parent = self

    def __hash__(self) -> int:
        return hash(id(self))

    def __repr__(self) -> str:
        children = ", ".join(str(child) for child in self.children)
        return f"Tree({self.label!r}, children=[{children}])"

    def __lt__(self, other: Tree) -> bool:
        return self.label < other.label

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Tree):
            return NotImplemented
        return self.label == other.label and sorted(self.children) == sorted(other.children)

    def find(self, label: str) -> Optional[Tree]:
        """Recursively search for a node by label."""
        if self.label == label:
            return self
        for child in self.children:
            if found := child.find(label):
                return found
        return None

    def adjacent(self) -> Generator[Tree, None, None]:
        """Return all adjacent nodes (parent + children)."""
        if self.parent:
            yield self.parent
        yield from self.children

    def children_from(self, parent: Optional[str]) -> list[Tree]:
        """Return children of `self`, assuming parent."""
        return [
            Tree(child.label, child.children_from(self.label))
            for child in self.adjacent()
            if child.label != parent
        ]

    def from_pov(self, from_node: str) -> Tree:
        """Build a Tree from a specific POV."""
        if not (start := self.find(from_node)):
            raise ValueError("Tree could not be reoriented")
        return Tree(start.label, start.children_from(None))

    def path_to(self, from_node: str, to_node: str) -> list[str]:
        """Return the path between two nodes."""
        if not (start := self.find(from_node)):
            raise ValueError("Tree could not be reoriented")

        path = {start.label: [start.label]}

        # Explore all nodes breadth first begining with from_node until to_node.
        to_explore = {start}
        while to_explore:
            # Explore another node.
            current = to_explore.pop()
            # If this is the destination, return the path to here.
            if current.label == to_node:
                return path[current.label]
            for neighbor in current.adjacent():
                if neighbor.label in path:
                    continue
                # Update the path to this node and add it to be explored.
                path[neighbor.label] = path[current.label] + [neighbor.label]
                to_explore.add(neighbor)

        raise ValueError("No path found")

File: go/pov/pov.go
package pov

import (
	"errors"
)

// Tree holds a tree structure.
type Tree struct {
	val    string
	parent *Tree
	// adjacencies = parent + children
	adjacencies []*Tree
}

// New creates and returns a new Tree with the given root value and children.
func New(value string, children ...*Tree) *Tree {
	t := &Tree{value, nil, children}
	for _, c := range children {
		c.adjacencies = append(c.adjacencies, t)
		c.parent = t
	}
	return t
}

// findNode searches downwards (ignoring the parent) for a node.
func (tr *Tree) findNode(value string) (*Tree, error) {
	if tr.Value() == value {
		return tr, nil
	}
	for _, child := range tr.Children() {
		if found, err := child.findNode(value); err == nil {
			return found, nil
		}
	}
	return nil, errors.New("unable to find node")
}

// setParent recursively updates the parent of a tree.
func (tr *Tree) setParent(parent *Tree) {
	tr.parent = parent
	for _, child := range tr.Children() {
		child.setParent(tr)
	}
}

// Value returns the value at the root of a tree.
func (tr *Tree) Value() string {
	return tr.val
}

// Children returns a slice containing the children of a tree.
func (tr *Tree) Children() []*Tree {
	var children []*Tree
	for _, node := range tr.adjacencies {
		if node != tr.parent {
			children = append(children, node)
		}
	}
	return children
}

// FromPov returns the pov from the node specified in the argument.
func (tr *Tree) FromPov(from string) *Tree {
	root, err := tr.findNode(from)
	if err != nil {
		return nil
	}
	root.setParent(nil)
	return root
}

// PathTo returns the shortest path between two nodes in the tree.
// O(n^2) time.
func (tr *Tree) PathTo(from, to string) []string {
	// Validate the from and to exist in the tree.
	if _, err := tr.findNode(from); err != nil {
		return nil
	}
	if _, err := tr.findNode(to); err != nil {
		return nil
	}

	tr = tr.FromPov(from)
	out := []string{tr.Value()}

	// Walk the tree one step at a time towards the "to".
	for ; tr.Value() != to; out = append(out, tr.Value()) {
		for _, child := range tr.Children() {
			// When we find a child leading towards the target, repeat from that child.
			if _, err := child.findNode(to); err == nil {
				tr = child
				break
			}
		}

	}
	return out
}
